src/sqlfluff/core/parser/lexer.py

```diff
@@ -254,6 +255,242 @@ def search(self, forward_string: str) -> Optional[Tuple[int, int]]:
+            Dedent(
+                is_template=True,
+                pos_marker=PositionMarker.from_point(
+                    last_source_slice.stop,
+                    templated_idx,
+                    templated_file,
+                ),
+            ),
+            TemplateLoop(
+                pos_marker=PositionMarker.from_point(
+                    last_source_slice.stop,
+                    templated_idx,
+                    templated_file,
+                )
+            ),
+            Indent(
+                is_template=True,
+                pos_marker=PositionMarker.from_point(
+                    last_source_slice.stop,
+                    templated_idx,
+                    templated_file,
+                ),
+            ),
+        ]
+    else:
+        return []
+
+
+def _generate_placeholder_segments(
+    source_slice: slice,
+    # NOTE: no last_source_slice implies start of file.
+    last_source_slice: Optional[slice],
+    template_slice: slice,
+    source_only_slices: List[RawFileSlice],
+    templated_file: TemplatedFile,
+    add_indents: bool,
+) -> Tuple[List[RawSegment], slice]:
+    """Generate any template placeholder segments.
+
+    The input source_slice, will potentially include not just
+    elements from the templated file, but also sections of
+    template code which no longer appear in the rendered file.
+
+    This code extracts them, adjusts the source_slice and generates
+    appropriate template segments to insert.
+
+    NOTE: For reasons which aren't well documented, any literal
+    elements will come at the end of the source_slice, and any
+    source only elements will usually come at the start. This
+    code takes advantage of that.
+    """
+    so_slices = []
+    # First check whether we've got any relevant source only slices for
+    # this position in the file.
+    if last_source_slice != source_slice:
+        for source_only_slice in source_only_slices:
+            # If it's later in the source, stop looking. Any later
+            # ones *also* won't match.
+            if source_only_slice.source_idx >= source_slice.stop:
+                break
+            elif source_only_slice.source_idx >= source_slice.start:
+                so_slices.append(source_only_slice)
+
+    # No relevant source only slices in this instance. Return
+    if not so_slices:
+        return [], source_slice
+
+    lexer_logger.debug("    Collected Source Only Slices")
+    for so_slice in so_slices:
+        lexer_logger.debug("       %s", so_slice)
+
+    # Calculate some things which will be useful
+    templ_str = templated_file.templated_str[template_slice]
+    source_str = templated_file.source_str[source_slice]
+
+    # For reasons which aren't entirely clear right now, if there is
+    # an included literal, it will always be at the end. Let's see if it's
+    # there.
+    if source_str.endswith(templ_str):
+        existing_len = len(templ_str)
+    else:
+        existing_len = 0
```

Comment:
I wonder if this logic is still needed. It potentially seems like a workaround from before `JinjaTracer`.

Reply:
It's still necessary for jinja tags which consume whitespace (like `{%- ... -%}`). Otherwise I'd agree.

Reply:
Tricky!